src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,175 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexInBounds(index);
+
+        if (size == index) {
+            linkLast(value);
+        } else {
+            Node<T> nodeIndex = getNodeIndex(index);
+            Node<T> prevNode = nodeIndex.prev;
+            Node<T> newNode = new Node<>(prevNode, value, nodeIndex);
+            nodeIndex.prev = newNode;
+
+            if (prevNode == null) {
+                head = newNode;
+            } else {
+                prevNode.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexInBounds(index);
+        checkIndexAtSizeRange(index);
+        return getNodeIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexInBounds(index);
+        checkIndexAtSizeRange(index);
+        Node<T> nodeIndex = getNodeIndex(index);
+        T changedNode = nodeIndex.value;
+        nodeIndex.value = value;
+        return changedNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexInBounds(index);
+        checkIndexAtSizeRange(index);
+        return unlink(getNodeIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> nodeIndex = head; nodeIndex != null; nodeIndex = nodeIndex.next) {
+                if (nodeIndex.value == null) {
+                    unlink(nodeIndex);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> nodeIndex = head; nodeIndex != null; nodeIndex = nodeIndex.next) {
+                if (object.equals(nodeIndex.value)) {
+                    unlink(nodeIndex);
+                    return true;
+                }
+            }
+        }
```

Comment:
we may perform one check for both null and not null cases
```suggestion
        for (Node<T> nodeIndex = head; nodeIndex != null; nodeIndex = nodeIndex.next) {
            if (object == nodeIndex.value || object != null && object.equals(nodeIndex.value)) {
                unlink(nodeIndex);
                return true;
            }
        }
```